astropy/wcs/wcsapi/high_level_api.py

```diff
@@ -117,116 +117,196 @@ def world_to_array_index(self, *world_objects):
-                    if 'frame' in kwargs:
-                        objects[key] = w.transform_to(kwargs['frame'])
-                    else:
-                        objects[key] = w
-                else:
-                    objects[key] = klass_gen(w, *args, **kwargs)
+def values_to_high_level_objects(*world_values, low_level_wcs):
+    """
+    Convert low level values into high level objects.
 
-        # We now extract the attributes needed for the world values
-        world = []
-        for key, _, attr in components:
-            if callable(attr):
-                world.append(attr(objects[key]))
-            else:
-                world.append(rec_getattr(objects[key], attr))
+    This function uses the information in ``wcs.world_axis_object_classes`` and
+    ``wcs.world_axis_object_components`` to convert the high level objects
+    (such as `~.SkyCoord`) to low level "values" `~.Quantity` objects.
+
+    This is used in `.HighLevelWCSMixin.pixel_to_world`, but provided as a
+    separate function for use in other places where needed.
+
+    Parameters
+    ----------
+    *world_values: object
+        Low level, "values" representations of the world coordinates.
+
+    low_level_wcs: `.BaseLowLevelWCS`
+        The WCS object to use to interpret the coordinates.
+    """
+    # Cache the classes and components since this may be expensive
+    components = low_level_wcs.world_axis_object_components
+    classes = low_level_wcs.world_axis_object_classes
+
+    # Deserialize classes
+    if low_level_wcs.serialized_classes:
+        classes_new = {}
+        for key, value in classes.items():
+            classes_new[key] = deserialize_class(value, construct=False)
+        classes = classes_new
+
+    args = defaultdict(list)
+    kwargs = defaultdict(dict)
+
+    for i, (key, attr, _) in enumerate(components):
+        if isinstance(attr, str):
+            kwargs[key][attr] = world_values[i]
+        else:
+            while attr > len(args[key]) - 1:
+                args[key].append(None)
+            args[key][attr] = world_values[i]
+
+    result = []
+
+    for key in default_order(components):
+        klass, ar, kw, *rest = classes[key]
+        if len(rest) == 0:
+            klass_gen = klass
+        elif len(rest) == 1:
+            klass_gen = rest[0]
+        else:
+            raise ValueError("Tuples in world_axis_object_classes should have length 3 or 4")
+        result.append(klass_gen(*args[key], *ar, **kwargs[key], **kw))
+
+    return result
+
+
+class HighLevelWCSMixin(BaseHighLevelWCS):
+    """
+    Mix-in class that automatically provides the high-level WCS API for the
+    low-level WCS object given by the `~HighLevelWCSMixin.low_level_wcs`
+    property.
+    """
+
+    @property
+    def low_level_wcs(self):
+        return self
+
+    def world_to_pixel(self, *world_objects):
+
+        world = high_level_objects_to_values(*world_objects, low_level_wcs=self.low_level_wcs)
```

Comment:
Minor: I think this should maybe be called 'world_values' - same comment a few times below